larq_zoo/literature/real_to_bin_nets.py

```diff
@@ -0,0 +1,494 @@
+        which has two convolutions.
+
+        Channel scaling follows Figure 1 (Right).
+        """
+
+        in_channels = x.shape[-1]
+        out_channels = in_channels * 2 if downsample else in_channels
+
+        # Shortcut, which gets downsampled if necessary
+        shortcut_add = self._shortcut_connection(x, name, in_channels, out_channels)
+
+        # Batch Normalization
+        conv_input = tf.keras.layers.BatchNormalization(
+            momentum=self.momentum, name=f"{name}_batch_norm"
+        )(x)
+
+        # Convolution
+        conv_output = lq.layers.QuantConv2D(
+            out_channels,
+            3,
+            strides=2 if downsample else 1,
+            padding="same",
+            input_quantizer=self.input_quantizer,
+            kernel_quantizer=self.kernel_quantizer,
+            kernel_initializer=self.kernel_initializer,
+            use_bias=False,
+            name=f"{name}_conv2d",
+        )(conv_input)
+
+        # binary convolution rescaling
+        x = self._scale_binary_conv_output(conv_input, conv_output, name)
+
+        # PReLU activation
+        x = tf.keras.layers.PReLU(shared_axes=[1, 2], name=f"{name}_prelu")(x)
+
+        # Skip connection
+        return tf.keras.layers.Add(name=f"{name}_skip_add")([x, shortcut_add])
+
+    def _block(self, x, downsample=False, name=""):
+        """Full binary block from Figure 1 (Left) of Matrinez et al. (2019)."""
+
+        x = self._half_binary_block(x, downsample=downsample, name=f"{name}a")
+        x = self._half_binary_block(x, downsample=False, name=f"{name}b")
+
+        # Add explicit name to the block output for attention matching (Section 4.2 of Martinez et al.)
+        return tf.keras.layers.Lambda(lambda x: x, name=f"{name}_out")(x)
+
+    def _get_weights_path(self):
+        if (
+            not self.kernel_quantizer == "ste_sign"
+            and self.input_quantizer == "ste_sign"
+        ):
+            raise NotImplementedError(
+                f"{self.model_name} only has stored weights for the BNN variant"
+            )
+        if self.include_top:
+            weights_path = utils.download_pretrained_model(
+                model="baseline_bnn",
+                version="v0.1.0",
+                file="r2b_strong_baseline_bnn_weights.h5",
+                file_hash="",  # TODO
+            )
+        else:
+            weights_path = utils.download_pretrained_model(
+                model="baseline_bnn",
+                version="v0.1.0",
+                file="r2b_strong_baseline_bnn_weights_notop.h5",
+                file_hash="",  # TODO
+            )
+        return weights_path
+
+
+@factory
+class RealToBinNetFactory(StrongBaselineNetFactory):
+    def _scale_binary_conv_output(
+        self, conv_input: tf.Tensor, conv_output: tf.Tensor, name: str
+    ):
+        """Data-dependent (squeeze and excite style) scaling of the convolution output as described in
+        Section 4.3 of Martinez at. al.
```

Comment:
```suggestion
        """Squeeze-and-Excite-style scaling.
        
        Data-dependent scaling of the convolution output, as described in Section 4.3
        of Martinez at. al.
```

Reply:
I changed this to:

```
        """Data-dependent convolution scaling.

        Scales the output of the convolution in the (squeeze-and-excite
        style) data-dependent way described in Section 4.3 of Martinez at. al.
        """
```
In the paper the link to squeeze and excite is not made very explicitly. They introduce it as their novel 'data-dependent scaling trick'. So if someone is looking for this magical data dependent scaling trick I think that should be the headline.

Reply:
Nice, makes sense.